L1Trigger/Phase2L1ParticleFlow/src/CaloClusterer.cc

```diff
@@ -0,0 +1,640 @@
+    iphi -= (iphi % 2);
+  iphi += 1;
+  //if (!valid_ieta_iphi(ieta,iphi)) {
+  //    printf("Error in finding cell for eta %+7.4f phi %+7.4f, got ieta = %+3d iphi %2d which is not valid\n",
+  //        eta, phi, ieta, iphi);
+  //}
+  assert(valid_ieta_iphi(ieta, iphi));
+  int icell = ifind_cell(ieta, iphi);
+  assert(icell != -1);
+
+  //if (std::abs(eta - eta_[icell]) > 0.501*etaWidth_[icell] || std::abs(deltaPhi(phi, phi_[icell])) > 0.501*phiWidth_[icell]) {
+  //    printf("Mismatch in finding cell for eta %+7.4f phi %+7.4f, got ieta = %+3d iphi %2d which has eta %+7.4f +- %.4f phi %+7.4f +- %.4f ; deta = %+7.4f dphi = %+7.4f\n",
+  //        eta, phi, ieta, iphi, eta_[icell], etaWidth_[icell], phi_[icell], phiWidth_[icell], eta - eta_[icell], deltaPhi(phi, phi_[icell]));
+  //}
+  //assert(std::abs(eta - eta_[icell]) <= 0.5*etaWidth_[icell]);
+  //assert(std::abs(deltaPhi(phi, phi_[icell])) <= 0.5*phiWidth_[icell]);
+  return icell;
+}
+
+int l1tpf_calo::Phase1GridBase::imove(int ieta, int iphi, int deta, int dphi) {
+  int ie = ieta, iph = iphi;
+  switch (deta) {
+    case -1:
+      ie = (ie == -nEta_ ? 0 : (ie == +1 ? -1 : ie - 1));
+      break;
+    case +1:
+      ie = (ie == +nEta_ ? 0 : (ie == -1 ? +1 : ie + 1));
+      break;
+    case 0:
+      break;
+    default:
+      assert(false);
+  };
+  if (ie == 0)
+    return -1;
+  switch (dphi) {
+    case -1:
+      iph = (iph == 1 ? nPhi_ : iph - 1);
+      break;
+    case +1:
+      iph = (iph == nPhi_ ? 1 : iph + 1);
+      break;
+    case 0:
+      break;
+    default:
+      assert(false);
+  };
+  if (!valid_ieta_iphi(ie, iph))
+    return -1;
+  int icell = ifind_cell(ie, iph);
+  assert(!(ie == ieta && iph == iphi));
+  assert(icell != -1);
+  assert(icell != ifind_cell(ieta, iphi));
+  return icell;
+}
+
+const l1tpf_calo::Grid *l1tpf_calo::getGrid(const std::string &type) {
+  static Phase1Grid _phase1Grid;
+  static Phase2Grid _phase2Grid;
+  if (type == "phase1")
+    return &_phase1Grid;
+  else if (type == "phase2")
+    return &_phase2Grid;
+  else
+    throw cms::Exception("Configuration") << "Unsupported grid type '" << type << "'\n";
+}
+
+l1tpf_calo::SingleCaloClusterer::SingleCaloClusterer(const edm::ParameterSet &pset)
+    : grid_(getGrid(pset.getParameter<std::string>("grid"))),
+      rawet_(*grid_),
+      unclustered_(*grid_),
+      precluster_(*grid_),
+      clusterIndex_(*grid_),
+      cellKey_(*grid_),
+      clusters_(),
+      nullCluster_(),
+      zsEt_(pset.getParameter<double>("zsEt")),
+      seedEt_(pset.getParameter<double>("seedEt")),
+      minClusterEt_(pset.getParameter<double>("minClusterEt")),
+      minEtToGrow_(pset.existsAs<double>("minEtToGrow") ? pset.getParameter<double>("minEtToGrow") : -1),
+      energyWeightedPosition_(pset.getParameter<bool>("energyWeightedPosition")) {
+  std::string energyShareAlgo = pset.getParameter<std::string>("energyShareAlgo");
+  if (energyShareAlgo == "fractions")
```

Comment:
eventually a factory pattern could be used for these algorithms
(to reduce repeated and somewhat less efficient switch statements)

Reply:
are there any results from code profiling that show that switch statements are somewhat less efficient than a "factory pattern" ?

Reply:
I have observed this in some instances. With a long switch statement containing many cases of string comparisons (in particular, since these are slow), it can take a non-negligibly longer amount of time to reach the cases at the end. Since this is a constructor, it probably doesn't contribute significantly to the execution time. However, using a factory also reduces the number of places to edit when adding a new algorithm, which is a good practice in general.